.moban.cd/moban.yml

```diff
@@ -4,13 +4,17 @@ organisation: moremoban
 author: C. W.
 contact: wangc_2011@hotmail.com
 license: MIT
-version: 0.5.0
-current_version: 0.5.0
+version: 0.6.0
+current_version: 0.6.0
 release: 0.5.0
 branch: master
 master: index
-command_line_interface: "moban"
-entry_point: "moban.main:main"
+entry_points:
+  console_scripts:
+    - "moban = moban.main:main"
+  fs.opener:
+    - "pypi = moban.fs_openers:PypiFSOpener"
```

Comment:
I am now quite convinced that the `pypi` protocol can be an external library that is re-usable.  It could even avoid the `pip install` "bug" https://github.com/moremoban/moban/issues/278 by doing a `pip download --only-binary :all:` (wheel fetch only) of the name/version/etc, and then use `archivefs` on the fetched file which is a wheel and thus has a known internal structure.

e.g. `pypi://pypi-mobans-pkg==1.1/resources/foo/bar.txt`

Note this `pypi://` opener also could give access to git repos, as `pip download` can also fetch vcs links as long as they have a `setup.py` or `pyproject.toml` .

The url for that could be `pypi+git://github.com/mobans/pypi-mobans-pkg@1.1/resources/foo/bar.txt` and `pypi+svn://github.com/mobans/pypi-mobans-pkg@15/resources/foo/bar.txt`

(I am not sure about that `+` -- need to check if pyfilesystem2 supports that.)

This is extra reason to not use the protocol name `pypi:` now , as its meaning will change if we do implement the above (which would remove the need for `requires:`).

Reply:
Maybe the `+` in the protocol could be achieved using registering them all in the opener
`protocols = ['pypi', 'pypi+git', 'pypi+hg', 'pypi+svn', 'pypi+bzr']`

Reply:
beyond theory, this might just work, at least as a temporary solution for our current limited usage of git repos.

`python3 -m pip download -b ./build -d ./download --use-pep517 --no-deps --no-build-isolation --verbose 'git+https://github.com/moremoban/pypi-mobans-pkg'` fetches and uses setup.py to get the correct version

`python3 -m pip download -b ./build -d ./download --use-pep517 --no-deps --no-build-isolation --verbose 'git+https://github.com/moremoban/pypi-mobans'` fetches and ignores missing setup.py and missing pyproject.toml to create UNKNOWN-0.0.0.zip which has the desired contents, including git submodules.

The `pypi+...` handlers would need to look for the new .zip file, because pip could give it any name at all.

It isnt pretty, but it is a useful hidden feature we could include in the `pypi://` opener

Reply:
I think github repo should be handled differently because it has branch, ref parameters. Python package has version and folder path. Mixing both would lead to complex solution down the line?

Reply:
pip vcs urls have branch and ref support.

But this would only be a temporary solution until a real `git://` is built with the features needed.

Reply:
So can we have an internal and temporary git opener(for git://) for our use case. Then when better git:// handler appears, we could swap. In this way, downstream project are not impacted in the future.

Reply:
The problem with that will be if the real git fs opener uses a different fs url format.

Reply:
backreference added to https://github.com/moremoban/pypifs/issues/4 , as this is now beyond scope of this PR (and this repo)